src/main/java/com/exactpro/th2/act/impl/SubscriptionManagerImpl.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021-2021 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.act.impl;
+
+import static com.google.protobuf.TextFormat.shortDebugString;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.exactpro.th2.act.SubscriptionManager;
+import com.exactpro.th2.common.grpc.Direction;
+import com.exactpro.th2.common.grpc.MessageBatch;
+import com.exactpro.th2.common.schema.message.MessageListener;
+
+public class SubscriptionManagerImpl implements MessageListener<MessageBatch>, SubscriptionManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SubscriptionManagerImpl.class);
+    private final Map<Direction, List<MessageListener<MessageBatch>>> callbacks = new ConcurrentHashMap<>();
+
+    public SubscriptionManagerImpl() {
+        callbacks.put(Direction.FIRST, new CopyOnWriteArrayList<>());
+        callbacks.put(Direction.SECOND, new CopyOnWriteArrayList<>());
+    }
+
+    @Override
+    public void register(Direction direction, MessageListener<MessageBatch> listener) {
+        List<MessageListener<MessageBatch>> listeners = getMessageListeners(direction);
+        listeners.add(listener);
+    }
+
+    @Override
+    public boolean unregister(Direction direction, MessageListener<MessageBatch> listener) {
+        List<MessageListener<MessageBatch>> listeners = getMessageListeners(direction);
+        return listeners.remove(listener);
+    }
+
+    @Override
+    public void handler(String s, MessageBatch messageBatch) {
+        if (messageBatch.getMessagesCount() < 0) {
+            if (LOGGER.isErrorEnabled()) {
+                LOGGER.error("Empty batch received {}", shortDebugString(messageBatch));
+            }
+            return;
+        }
+        Direction direction = messageBatch.getMessages(0).getMetadata().getId().getDirection();
+        List<MessageListener<MessageBatch>> listeners = callbacks.get(direction);
+        if (listeners == null) {
+            if (LOGGER.isErrorEnabled()) {
+                LOGGER.error("Unsupported direction {}. Batch: {}", direction, shortDebugString(messageBatch));
+            }
+            return;
+        }
```

Comment:
Why is it not an error? It is an exceptional case when we received the unsupported direction and cannot route it anywhere

Reply:
What is mean under `unsupported direction`?
It is a fatal problem when the box isn't compatible with the environment.
It is a warning when the box works only with specific directions.

Reply:
We can rename it to `Unknown direction`) The point is that it is something critical but still, we cannot throw an exception here. It might be a mistake in a component that generated the received batch. So, we need to report it and continue working.
And there is not `fatal` method in the logger. So, the best choice is `error` I think